### ISSUE TITLE
Q4 GGUF loads successfully but generation outputs repeated BOS tokens only

### ISSUE THREAD

**Issue by ensemble-learning** [antirez/ds4#51]
# Q4 GGUF loads successfully but generation outputs repeated BOS tokens only

## Bug Description

`ds4` and `ds4-server` can successfully load the Q4 DeepSeek V4 Flash GGUF model with the Metal backend, initialize the graph, and serve API requests. However, actual generation produces only repeated BOS tokens:

```
<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>...
```

This happens both in the CLI and through the OpenAI-compatible `/v1/chat/completions` API.

## Repository / Commit

```
repo: git@github.com:antirez/ds4.git
commit: 8e7575b Fix JSON parser nesting DoS
```

Working tree appeared clean during testing.

## Environment

```
Machine: Apple Silicon Mac Studio
OS: macOS 26.4.1
Kernel: Darwin 25.4.0 RELEASE_ARM64_T6031
Architecture: arm64
RAM: 512 GB
Backend: Metal
```

## Model

Default model symlink:

```
ds4flash.gguf -> ./gguf/DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf
```

Model file size:

```
165102258845 bytes
~154G on disk
```

`./ds4 --inspect` succeeds:

```
model: DeepSeek V4 Flash
arch:  deepseek4
gguf:  v3, 58 metadata keys, 1328 tensors
layers: 43
attention: heads=64 kv_heads=1 head_dim=512 swa=128
indexer: heads=64 head_dim=128 top_k=512
experts: count=256 used=6 groups=0 groups_used=0
file size: 153.76 GiB
tensor bytes described by GGUF: 153.32 GiB
logical parameters: 284.33 B
tensor types:
  f32        492 tensors, 0.00 GiB
  f16        359 tensors, 2.04 GiB
  q8_0       345 tensors, 6.15 GiB
  q4_k       129 tensors, 145.12 GiB
  i32          3 tensors, 0.01 GiB
```

Metal initialization also appears successful:

```
ds4: Metal model views created in 4.536 ms, residency requested in 1350.597 ms, warmup 7.469 ms
ds4: Metal mapped mmaped model as 1 overlapping shared buffers
ds4: Metal backend initialized for graph diagnostics
```

### 1. Build

```
cd /Users/tcheng/soft/ds4/ds4

make
```

Result:

```
make: Nothing to be done for `all'.
```

### 2. Inspect model

```
./ds4 --inspect
```

This succeeds and prints a valid model summary.

### 3. Run CLI generation

```
./ds4 --nothink --temp 0 -n 16 -p 'Say hello in one short sentence.'
```

### 4. Optional: run server

 local service use port `8001`:

```
./ds4-server \

  --port 8001 \

  --ctx 32768 \

  --kv-disk-dir /tmp/ds4-kv \

  --kv-disk-space-mb 1024 \

  --trace /tmp/ds4-server-8001.trace
```

Verify models endpoint:

```
curl -fsS http://127.0.0.1:8001/v1/models
```

This succeeds and returns `deepseek-v4-flash`.

### 5. Run API generation

```
curl -fsS http://127.0.0.1:8001/v1/chat/completions \

  -H 'Content-Type: application/json' \

  -d '{

    "model":"deepseek-chat",

    "messages":[{"role":"user","content":"Say hello in one short sentence."}],

    "max_tokens":8,

    "temperature":0

  }'
```

## Expected Behavior

The model should generate a normal short assistant response, for example:

```
Hello, nice to meet you!
```

or any semantically valid natural-language completion.

## Actual Behavior

CLI output:

```
<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>
```

CLI performance line:

```
ds4: prefill: 65.80 t/s, generation: 37.51 t/s
```

API response:

```
{

  "id": "chatcmpl-1",

  "object": "chat.completion",

  "model": "deepseek-chat",

  "choices": [

    {

      "index": 0,

      "message": {

        "role": "assistant",

        "content": "<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>"

      },

      "finish_reason": "length"

    }

  ],

  "usage": {

    "prompt_tokens": 11,

    "completion_tokens": 8,

    "total_tokens": 19

  }

}
```

## Notes

-   The model loads successfully.
-   Metal backend initialization succeeds.
-   Prefill runs through all 43 layers.
-   Server API works structurally.
-   The failure appears to be in inference correctness, logits, sampling, tokenizer mapping, or Q4 GGUF compatibility.
-   The symptom is deterministic with `--temp 0`: every generated token is BOS.

## Additional Context

The server process used for this test has been stopped after reproduction. Port `8001` is no longer listening.

Potential areas to inspect:

-   logits after first decode step
-   whether logits are all `-inf` except token `0`
-   sampler behavior under greedy decode
-   BOS / special-token handling
-   Q4 tensor loading / dequant path
-   output projection / final norm path
-   tokenizer id mapping for `<｜begin▁of▁sentence｜>`

**Comment by antirez** [comment]
Is this an M3 Ultra? Thanks

**Comment by antirez** [comment]
I noticed the GGUF file size does not match... Likely corrupted file.

**Comment by ensemble-learning** [comment]
Yes, it is M3 Ultra. I double check the file size, which appears to be the same in hg as 165,102,258,845 bytes = 165.10 GB ≈ 153.76 GiB.
Sorry for the confusion by previously reporing the GiB unit.

**Comment by ensemble-learning** [comment]
I am downloading the q2 version to see if it works.

**Comment by kucats** [comment]
I had same issue and produced only

<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>...

gguf file was correct size but had wrong sha256 - redownload did fix it. Thought I'll get more obvious error if one of the file corrupted.

**Comment by ensemble-learning** [comment]
@kucats thanks. still downloading. It is quite big. Hopefully, this time will be all right.
